File: python/solveCCH.py
import math
import cmath
import numpy as np
from matplotlib import pyplot as plt


class SOLVECCH:
    def __init__(self):
        self.a = None
        self.b = None
        self.c = None
        self.y_o = None
        self.y_prime = None
        self.X = np.linspace(-1 * np.pi, np.pi, 100)

    def get_inputs(self):
        """
        Gets the user inputs for the coefficients and initial conditions
        a: The coefficient a
        b: The coefficient b
        c: The coefficient c
        y_o: Initial condition when y(0) is equal to a value
        y_prime: Initial condition when y'(0) is equal to a value
        """
        self.a = float(input("Enter the value of a: "))
        self.b = float(input("Enter the value of b: "))
        self.c = float(input("Enter the value of c: "))
        self.y_o = float(input("Enter the value of y(0): "))
        self.y_prime = float(input("Enter the value of y`(0): "))

    def plot_solution(self, eq: np.array) -> None:
        """
        Plots the equation in the solution
        :param eq: The characteristic equation
        """
        fig = plt.figure(figsize=(14, 8))
        plt.plot(self.X, eq, 'b', label='Data points')
        plt.legend()
        plt.grid(True, linestyle=':')
        plt.xlim([-6, 6])
        plt.ylim([-4, 4])
        plt.title('CCH')
        plt.xlabel('x-axis')
        plt.ylabel('y-axis')
        plt.show()

    def solve_cch(self) -> bool:
        """
        This function solve the quadratic equations the produces the output equation
        :rtype: bool

        :return: The status of the function
        """
        discriminant = self.b ** 2 - 4 * self.a * self.c
        if discriminant > 0:
            print("The roots are real")
            r1 = (-self.b + math.sqrt((self.b * self.b) - (4 * self.a * self.c))) / (2 * self.a)
            r2 = (-self.b - math.sqrt((self.b * self.b) - (4 * self.a * self.c))) / (2 * self.a)
            A = np.array([[1, 1], [r1, r2]])
            B = np.array([self.y_o, self.y_prime])
            C = np.linalg.solve(A, B)
            eq = round(C[0], 4) * np.exp(r1 * self.X) + round(C[1], 4) * np.exp(r2 * self.X)
            print("Equation is: y = {}e^{}t + {}e^{}t".format(round(C[0], 4), r1, round(C[1], 4), r2))
            self.plot_solution(eq=eq)

        elif discriminant == 0:
            print("The roots are repeated")
            r1 = (-self.b + math.sqrt((self.b * self.b) - (4 * self.a * self.c))) / (2 * self.a)
            eq = round(self.y_o, 4) * np.exp(r1 * self.X) + round((self.y_prime - r1 * self.y_o), 4) * self.X * np.exp(
                r1 * self.X)
            print("Equation is: y = {}e^{}t + {}te^{}t".format(round(self.y_o, 4), r1,
                                                               round((self.y_prime - r1 * self.y_o), 4), r1))
            self.plot_solution(eq=eq)

        elif discriminant < 0:
            print("The roots are imaginary")
            r1 = (-self.b + cmath.sqrt((self.b * self.b) - (4 * self.a * self.c))) / (2 * self.a)
            r2 = (-self.b - cmath.sqrt((self.b * self.b) - (4 * self.a * self.c))) / (2 * self.a)
            A = np.array([[1, 0], [r1.real, r1.imag]])
            B = np.array([self.y_o, self.y_prime])
            C = np.linalg.solve(A, B)
            eq = np.exp(r1.real * self.X) * (
                    round(C[0], 4) * np.cos(r1.imag * self.X) + round(C[1], 4) * np.sin(-1 * r2.imag * self.X))
            print("Equation is: y = e^{}t ({} cos({}t) + {} sin({}t))".format(r1.real, round(C[0], 4), r1.imag,
                                                                              round(C[1], 4), -1 * r2.imag))
            self.plot_solution(eq=eq)

        return True

    def main(self):
        self.get_inputs()
        self.solve_cch()


if __name__ == "__main__":
    my_solve = SOLVECCH()
    my_solve.main()

File: matlab/solveCCH.m
% Aurthor:  Rodney Osodo
% Course:   Bsc. Mechatronic Enhineering

function solveCCH
    % main function
    [a, b, c, y_o, y_prime] = get_inputs();
    equation1 = solve_cch(a, b, c, y_o, y_prime);
    equation2 = solveEquationWithDSolve(a,b,c, y_o, y_prime);
    ez1 = ezplot(str2sym(equation1) ,[-2 * pi,2 * pi]);
    hold on;
    ez2 = ezplot(equation2, [-2 * pi, 2 * pi]);
end

function eq = solve_cch(a, b, c, y_o, y_prime)
    discriminant = b .* b - 4  .*  a  .*  c;
    if (discriminant > 0)
        % 0.8*exp(3*t) + 0.2*exp(-2*t)
        fprintf("The roots are real\n");
        r1 = (-b + sqrt((b .* b) - (4 .* a .* c))) ./ (2 .* a);
        r2 = (-b - sqrt((b .* b) - (4 .* a .* c))) ./ (2 .* a);
        A = [1 1;r1 r2];
        B = [y_o; y_prime];
        C = linsolve(A,B);
        eq = string(C(1)) + "*exp(" + string(r1) + "*t)" + " + " + string(C(2)) + "*exp(" + string(r2) + "*t)";
        disp(eq);
    elseif (discriminant == 0)
        % 12*exp(2*t)+-27*t*exp(2*t)
        fprintf("The roots are repeated\n");
        r1 = (-b + sqrt((b .* b) - (4 .* a .* c))) ./ (2 .* a);
        eq = string(round(y_o, 4)) + "*exp(" + string(r1) + "*t)" + "+" + string(round((y_prime - r1 .* y_o), 4)) + "*t*exp(" + string(r1) + "*t)";
        disp(eq);
    elseif (discriminant < 0)
        % 1*cos(2*t)*exp(5*t) - 1*sin(2*t)*exp(5*t)
        fprintf("The roots are imaginary\n");
        r1 = (-b + sqrt((b .* b) - (4 .* a .* c))) ./ (2 .* a);
        r2 = (-b - sqrt((b .* b) - (4 .* a .* c))) ./ (2 .* a);
        A = [1 0;real(r1) imag(r1)];
        B = [y_o; y_prime];
        C = linsolve(A,B);
        eq = string(round(C(1), 4)) + "*cos(" + string(imag(r1)) + "*t)" + "*exp(" + string(real(r1)) + "*t)" + " - " + string(round(C(1), 4)) + "*sin(" + string(-1  .*  imag(r2)) + "*t)" + "*exp(" + string(real(r1)) + "*t)";
        disp(eq);
    end
end

function dSolveSolution = solveEquationWithDSolve(a, b, c, y1, y2)
    % solves using dsolve
    syms y(t)
    Dy = diff(y, t);
    ode_eqn = a * diff(y, t, 2) + b * diff(y, t) + c * y == 0;
    cond = [y(0) == y1, Dy(0) == y2];
    dSolveSolution = dsolve(ode_eqn, cond);
    disp('DSolve Solution');
    disp(dSolveSolution);
end

function [a, b, c, y_o, y_prime] = get_inputs()
    % gets user input
    syms a b c y(t) dy(t)
    genericEqn = a*diff(y,2) + b *diff(y) + c*y == 0;
    disp("Enter an equation of the form:");
    disp(genericEqn);
    a = input('Enter the value of a: ');
    b = input('Enter the value of b: ');
    c = input('Enter the value of c: ');
    y_o = input('Enter the value of y(0): ');
    y_prime = input('Enter the value of y`(0): ');
end